### PR TITLE
Generate can return cross-attention weights too

### LINKED ISSUE #10335: Return cross-attention weights in generation function
# 🚀 Feature request

With v4.2.0 release, generation can now return encoder and decoder self-attention weights but it still doesn't return cross-attention weights. These weights are already computed and returned in model `forward` method, and just need to be returned in `generate` method.
## Motivation

Visualizing cross-attention weights is useful for many applications such as token-alignment.

## Your contribution

I can submit a PR.

### PR DESCRIPTION
**Pull request by Mehrad0711**
# What does this PR do?

Fixes #10335 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

@patrickvonplaten

### CODE REVIEW

**Comment by Mehrad0711** [comment]
I tried to keep code changes as minimum as possible. Thus, I avoided adding another argument for returning cross-attention weights and used `output_attentions` to check instead. 

Also in docstrings, for `decoder_attentions` the shape is mentioned as `(batch_size*num_return_sequences, num_heads, generated_length,
            sequence_length)`, although it seems that the shape is always `(batch_size*num_return_sequences, num_heads, 1, 1)`. `generated_length` is 1 as the code returns tuples per generated token and `sequence_length` is 1 since it's decoder self-attentions.
Similarly for `cross_attentions` the shape should be `(batch_size, num_heads, 1, input_sequence_length)`. Are there any examples where this is not True?

Please let me know what you think. Thanks!

File: src/transformers/generation_utils.py [hunk 1]
```diff
@@ -1968,17 +1999,18 @@ def beam_sample(
             if not output_scores:
                 sequence_outputs["sequence_scores"] = None
             if self.config.is_encoder_decoder:
-                return BeamSearchEncoderDecoderOutput(
+                return BeamSampleEncoderDecoderOutput(
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:2002]: great catch!

**Comment by patrickvonplaten** [comment]
> Hi, I tried to keep the code changes to a minimum. Thus, I avoided adding another argument for returning cross-attention weights and used `output_attentions` to check instead.
> Also in docstrings, for `decoder_attentions` the shape is mentioned as `(batch_size*num_return_sequences, num_heads, generated_length, sequence_length)`, although it seems that the shape is always `(batch_size*num_return_sequences, num_heads, 1, 1)`. `generated_length` is 1 as the code returns tuples per generated token and `sequence_length` is 1 since it's decoder self-attentions.
> Similarly for `cross_attentions` the shape should be `(batch_size, num_heads, 1, input_sequence_length)`. Are there any examples where this is not True?
> 
> Please let me know what you think. Thanks!

Regarding the `decoder_attentions` shape -> it's just the first attentions that is of shape `batch_size, num_heads, 1, 1`. Then it goes up to `..., 2, 2`, `..., 3, 3` etc. However if `use_cache` is enabled, then the shape is `...1, 1`, `...1, 2`, `....1, 3`.

Check this example:

```python
from transformers import BartForConditionalGeneration
import torch

bart = BartForConditionalGeneration.from_pretrained("facebook/bart-large-cnn")
bart.config.max_length = 5 
bart.config.num_beams = 1

outputs_use_cache = bart.generate(torch.tensor([ 10 * [0]]), return_dict_in_generate=True, output_attentions=True)
outputs_no_cache = bart.generate(torch.tensor([ 10 * [0]]), return_dict_in_generate=True, output_attentions=True, use_cache=False)

# outputs_no_cache.decoder_attentions[1][0].shape gives (1, 16, 2, 2)
# outputs_use_cache.decoder_attentions[1][0].shape gives (1, 16, 1, 2)
```

**Review by patrickvonplaten — APPROVED** [review]
Great work!

**Comment by Mehrad0711** [comment]
> Regarding the decoder_attentions shape -> it's just the first attentions that is of shape batch_size, num_heads, 1, 1. Then it goes up to ..., 2, 2, ..., 3, 3 etc. However if use_cache is enabled, then the shape is ...1, 1, ...1, 2, ....1, 3.

Thanks for the example! It makes sense now.

**Review by patil-suraj — APPROVED** [review]
LGTM, Thanks a lot for your work.

**Comment by xwuShirley** [comment]
Thanks for the great work. But I am still a bit confused about getting the cross-attention or encode-decoder attention if I am not mistaken the cross attention. Any hint would be great. Maybe it would be like 

# outputs_no_cache.cross_attentions[1][0]
# outputs_use_cache.cross_attentions[1][0]

Thanks! @Mehrad0711

**Comment by patil-suraj** [comment]
hi @xwuShirley 

Yes, you are right. You can access the cross attentions using 
`outputs. cross_attentions`

**Pull request merged**